pkg/resource/deploy/deployment_executor.go

```diff
@@ -662,6 +662,12 @@ func (ex *deploymentExecutor) rebuildBaseState(resourceToStep map[*resource.Stat
 			new.Dependencies = deps
 		}
 
+		if new.Parent != "" {
```

Comment:
This function before the changes is pretty tough to follow and I'm having a hard time trusting it does the right thing,  with the addition increasing the conceptual burden further.

- is there any way to add the new code in a separate loop and function dedicated to cleaning up parents and with its own docstring?

- if we keep this code in here, could we comment a bit on what's it doing and why, and add a comment to the docstring of the function that it also takes care of parent links

- why `olds[new.Parent]` vs `referenceable[d]`?  
 
- it is concerning that both `olds` and `referenceable` are constructed incrementally over `range ex.deployment.prev.Resources` that uses the top-sort assumption for dependencies; does Parent patch up also need the top-sort assumption or is it somehow compatible with it because a Parent link is also a dependency link?

Reply:
I'm going to address each point in turn:
- Yes, and it reads much cleaner. See https://github.com/pulumi/pulumi/pull/10073/commits/c2cd1313170be05230f8a0b314e8e1a76fc51739.
- I added a docstring to the new function I introduced.
- Because olds doesn't include resources with pending deletes. https://github.com/pulumi/pulumi/blob/c2cd1313170be05230f8a0b314e8e1a76fc51739/pkg/resource/deploy/deployment_executor.go#L669-L672
- Parent is part of the requirement of topological sort, and we do rely on this property.